torch/nn/parallel/distributed.py

```diff
@@ -714,7 +715,20 @@ def _recursive_to(self, inputs, target_gpu):
         """
         def to_map(obj):
             if isinstance(obj, torch.Tensor):
-                return (obj.to(target_gpu), )
+                # Perform CPU -> GPU copies in a background stream. This code is
+                # motivated from similar logic in torch/nn/parallel/_functions.py
+                stream = _get_stream(target_gpu)
```

Comment:
Shouldn't `torch.cuda.Stream()` give you a new stream from the pool? (https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/Stream.cpp#L39)

Reply:
@pritamdamania87 To simulate the loggic in `scatter` (see: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L85), I used `_get_stream` which is what scatter does. It looks like `_get_stream` calls into torch.cuda.Stream(), but it caches it for each device for re-use, instead of allocating a fresh stream everytime: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L112

Reply:
Hmm I'm wondering what is the purpose of the caching in `_get_stream` since `torch.cuda.Stream()` also returns a stream from a pool and doesn't allocate a new one each time.

Reply:
Are we guaranteed that torch.cuda.Stream() will give us the same stream everytime for a given device, though? I'm not sure if that's necessary here but the behavior with `_get_stream()` makes that the case. 

The original code of caching the streams in `_get_stream` is from a few years ago: https://github.com/pytorch/pytorch/pull/955

Reply:
No, `torch.cuda.Stream()` won't give you the same stream every time. It is possible that at the time of #955 there was no stream pool, so caching the stream was a way to make sure a large number of streams was not created and destroyed.